### ISSUE TITLE
Invalid overload overlap detection with generic callable as argument

### ISSUE THREAD

**Issue by tkukushkin** [microsoft/pyright#3115]
**Describe the bug**
Invalid overload overlap detection with generic callable as argument. it's interesting that everything else works as intended.

**To Reproduce**
```python
from collections.abc import Callable
from typing import TypeVar, overload

T = TypeVar('T')

@overload
def foo(f: Callable[..., T]) -> Callable[..., T]: ...

@overload
def foo(f: Callable[..., T | None]) -> Callable[..., T | None]: ...  # error: Overload 2 for "foo" will never be used because its parameters overlap overload 1 (reportOverlappingOverload)

def foo(f: Callable[..., T | None]) -> Callable[..., T | None]: ...

def f1(x: int) -> str: ...
def f2(x: int) -> str | None: ...

reveal_type(foo(f1))  # information: Type of "foo(f1)" is "(...) -> str"
reveal_type(foo(f2))  # information: Type of "foo(f1)" is "(...) -> str | None"

```

**Expected behavior**
No `reportOverlappingOverload` errors.

**VS Code extension or command-line**
pyright 1.1.224

**Comment by hmc-cs-mdrissi** [comment]
One workaround you could do is you can specify the type argument to cached explicitly. So like,

```python
@cached[str](serializer=pickle_serializer)
def f3(x: int) -> str:
    ...


@cached[str](serializer=pickle_serializer)
def f4(x: int) -> str | None:
    ...
```

One caveat is that syntax is python 3.9+. If on python 3.8 or lower you can instead write,

```python
str_cached = cached[str]

@str_cached(serializer=pickle_serializer)
def f3(x: int) -> str:
    ...


@str_cached(serializer=pickle_serializer)
def f4(x: int) -> str | None:
    ...
```

**Comment by tkukushkin** [comment]
Didn't think of that option, thanks, it works! But then it is not clear what to do with the error:

```
error: Overload 2 for "foo" will never be used because its parameters overlap overload 1 (reportOverlappingOverload)
```

**Comment by tkukushkin** [comment]
I also have another interface to what the `cached` decorator does:

```python
@overload
def get_or_set(
        self,
        factory: Callable[[], T],
        serializer: Serializer[T],
) -> T: ...

@overload
def get_or_set(
        self,
        factory: Callable[[], T | None],
        serializer: Serializer[T],
) -> T | None: ...
```

In this case I can't specify type argument explicitly, maybe there are another workarounds for such case?

**Comment by tkukushkin** [comment]
I think there is bug in pyright. Pyright reports
```
error: Overload 2 for "foo" will never be used because its parameters overlap overload 1 (reportOverlappingOverload)
```
but it is not true, "overload 2" can be used in practice.

Code:

```python
import pickle
from collections.abc import Callable
from typing import Any, Generic, ParamSpec, Protocol, TypeVar, overload

P = ParamSpec('P')
T = TypeVar('T')

class Serializer(Protocol[T]):
    def dumps(self, value: T, /) -> bytes: ...
    def loads(self, value: bytes, /) -> T: ...

pickle_serializer: Serializer[Any] = pickle
str_serializer: Serializer[str] = ...

class cached(Generic[T]):
    def __init__(self, serializer: Serializer[T]) -> None: ...

    @overload
    def __call__(self, f: Callable[P, T]) -> Callable[P, T]: ...

    @overload
    def __call__(self, f: Callable[P, T | None]) -> Callable[P, T | None]: ...  #  error: Overload 2 for "__call__" will never be used because its parameters overlap overload 1

@cached[str](serializer=str_serializer)
def f1(x: int) -> str: ...

@cached[str](serializer=str_serializer)
def f2(x: int) -> str | None: ...

@cached[str](serializer=pickle_serializer)
def f3(x: int) -> str: ...

@cached[str](serializer=pickle_serializer)
def f4(x: int) -> str | None: ...

reveal_type(f1)  #  information: Type of "f1" is "(x: int) -> str"
reveal_type(f2)  # information: Type of "f2" is "(x: int) -> str | None"
reveal_type(f3)  #  information: Type of "f3" is "(x: int) -> str"
reveal_type(f4)  # information: Type of "f4" is "(x: int) -> str | None"
```

**Comment by erictraut** [comment]
This will be fixed in the next release.

**Comment by tkukushkin** [comment]
Can I ask you to attach to the message commits in which you fix bugs? It's very interesting to me to see changes in the code and especially test improvements.

**Comment by erictraut** [comment]
That doesn't work well for my normal workflow, but you can easily find the corresponding commit by looking at the time stamp and commit notes.

**Comment by hmc-cs-mdrissi** [comment]
```python
@overload
def get_or_set(
        self,
        factory: Callable[[], T],
        serializer: Serializer[T],
) -> T: ...

@overload
def get_or_set(
        self,
        factory: Callable[[], T | None],
        serializer: Serializer[T],
) -> T | None: ...
```

For this example, if your main goal is return type you can use bidirectional type inference and just tell it what it should be. So,

```python
x = get_or_set(...)
```

if it's picking wrong overload due to ambiguity you can do,

```python
x: expected_type = get_or_set(...)
```

The annotation can be used to help pick the overload.

If you have any follow up questions I'd ask in the discussions [part](https://github.com/microsoft/pyright/discussions).

**Comment by tkukushkin** [comment]
Sorry for off-topic. Thanks, this workaround also works and I noticed that pyright even without an explicit annotation right infers result type if `factory` argument passed to function before `serializer` (when its type is `Serializer[Any]` like `pickle_serializer `), it chooses more narrow type, I hope this behavior doesn't change over time. 

Also I noticed that mypy in this case chooses wider type and without explicit type annotation result type will be `Any`. But the good thing is that explicit annotation with the wrong type (e.g. `int` when the `factory` result type is `int | None`) leads to an error. I originally thought that since the "default" type is `Any`, mypy would just let me assign it to variable of any type, I didn't think type inference understood that. But the bad thing is that if the library user does not specify an explicit type, he can use the result as `Any` without any warning from mypy even in strict mode.

I have no more questions, thank you very much and sorry again for the off-topic.
